dask/dataframe/io/parquet.py

```diff
@@ -550,7 +554,7 @@ def _read_pyarrow(fs, fs_token, paths, columns=None, filters=None,
 
     all_columns = index_names + column_names
 
-    dtypes = _get_pyarrow_dtypes(schema)
+    dtypes = _get_pyarrow_dtypes(schema, categories)
     dtypes = {storage_name_mapping.get(k, k): v for k, v in dtypes.items()}
 
     meta = _meta_from_dtypes(all_columns, dtypes, index_names, column_index_names)
```

Comment:
Since we don't know the actual categories at graph build time, we need to make sure any categorical columns/indices are marked as *unknown* categories (see http://dask.pydata.org/en/latest/dataframe-design.html#categoricals). The fastparquet reader handles this here: https://github.com/dask/dask/pull/3177/files#diff-410a8a0995091d42fbbbc5a50efda3aeL295, but this looks incorrect for the no-data case, as `meta` is returned directly on compute and would then have the string `'__UNKNOWN_CATEGORIES__'` in its categories.

Ideally we'd have two objects:
- An empty frame/series that can be returned directly to the user on compute in the case of no data
- An empty frame/series to be used as metadata that properly sets unknown categories

These should ideally be handled using the same code path (perhaps move logic to `_meta_from_dtypes`). You might also find `dask.dataframe.utils.clear_known_categories` (sets all categories to be unknown) and `dask.dataframe.utils.strip_unknown_categories` (removes unknown categories flags) to be useful.

Reply:
Adjusted it hopefully in the correct way. Moving the code to `_meta_from_dtype` depends on the `pyarrow` path also supporting known categoricals. I'm not sure how soon I'll get to that.